lib/charms/loki_k8s/v0/log_proxy.py

```diff
@@ -321,12 +338,86 @@ def _create_directories(self) -> None:
         self._container.make_dir(path=WORKLOAD_CONFIG_DIR, make_parents=True)
 
     def _obtain_promtail(self, event) -> None:
-        if self._is_promtail_binary_in_workload():
+        """Obtain promtail binary from an attached resource or download it."""
+        if self._is_promtail_attached():
+            return
+
+        if not self._promtail_must_be_downloaded():
             return
 
         if self._download_promtail(event):
             with open(BINARY_PATH, "rb") as f:
                 self._container.push(WORKLOAD_BINARY_PATH, f, permissions=0o755, make_dirs=True)
+                logger.debug("Promtail binary file has been pushed to workload container.")
+
+    def _is_promtail_attached(self) -> bool:
+        """Checks whether Promtail binary is attached to the charm or not.
+
+        Returns:
+            a boolean representing whether Promtail binary is attached or not.
+        """
+        try:
+            resource_path = self._charm.model.resources.fetch("promtail-bin")
+        except ModelError:
+            return False
+
+        with open(resource_path, "rb") as f:
+            self._container.push(WORKLOAD_BINARY_PATH, f, permissions=0o755, make_dirs=True)
+
+        logger.info("Promtail binary file has been obtained from an attached resource.")
+        return True
+
+    def _promtail_must_be_downloaded(self) -> bool:
+        """Checks whether promtail binary must be downloaded or not.
+
+        Returns:
+            a boolean representing whether Promtail binary must be downloaded or not.
+        """
+        if not self._is_promtail_binary_in_workload():
+            return True
+        else:
+            self._get_promtail_bin_from_workload()
+
+        if not self._sha256sums_matches(BINARY_PATH, BINARY_SHA256SUM):
+            return True
+
+        logger.debug("Promtail binary file is already in the the workload container.")
+        return False
+
+    def _get_promtail_bin_from_workload(self) -> None:
+        """Gets the Promtail binary from the workload container into the charm container."""
+        binary = self._container.pull(WORKLOAD_BINARY_PATH, encoding=None).read()
+        with open(BINARY_PATH, "wb") as f:
+            f.write(binary)
+
+    def _sha256sums_matches(self, file_path: str, sha256sum: str) -> bool:
+        """Checks whether a file's sha256sum matches or not with an specific sha256sum.
+
+        Args:
+            file_path: A string representing the files' patch.
+            sha256sum: The sha256sum against which we want to verify.
+
+        Returns:
+            a boolean representing whether a file's sha256sum matches or not with
+            an specific sha256sum.
+        """
+        try:
+            with open(file_path, "rb") as f:
+                file_bytes = f.read()
```

Comment:
Probably not very relevant for our use case but see [comment on SO](https://stackoverflow.com/a/22058673/3516684) for buffering large files.